Adding shmem_malloc_with_hints interface (#259)
openshmem-org/specification: content/shmem_malloc_hints.tex

\end{Csynopsis}

\begin{apiarguments}
    \apiargument{IN}{size}{The size, in bytes, of a block to be
        allocated from the symmetric heap. This argument is of type \CTYPE{size\_t}}
    \apiargument{IN}{hints}{A bit array of hints provided by the user to the implementation}
\end{apiarguments}

\apidescription{

    The \FUNC{shmem\_malloc\_with\_hint} like \FUNC{shmem\_malloc} routine returns a pointer to a block of at least
    \VAR{size} bytes, which shall be suitably aligned so that it may be
    assigned to a pointer to any type of object.  This space is allocated from
    the symmetric heap (similar to \FUNC{shmem\_malloc}).  When \VAR{size} is zero, 
    the \FUNC{shmem\_malloc\_with\_hint} routine performs no action and returns a null pointer. 
    
    In addition to the \VAR{size} argument, the \VAR{hint} argument is provided by the user. 
    The \VAR{hint} could describe the trait of memory storage where memory is allocated or 
    it could describe the expected manner in which the \openshmem program may use the allocated memory.
    The valid storage trait hints are described in Table~\ref{traithints} and the valid usage hints 

>     are described in Table~\ref{usagehints}. Multiple hints are expressed as \CONST{OR} of \VAR{hints}.

Reviewer: @manjugv 
question regarding ```Multiple hints are expressed as OR of hints.```: how oshmem should react on case if supported less flags than requested? or supported all honts, but not at same buffer (and I believe this case - supported all but not at same buffer - is the most actual)?

it seems ```OR``` combination of ```traithints``` hints should be denied

Author: @hoopoepg That is correct. In the last sentence of this paragraph I say that. Some combination of hints does not make sense and some combination of hints might not be supported by a particular implementation. In either case, it can behave like shmem_malloc.

    and any subsequent \openshmem calls is undefined.
}

\apireturnvalues{
    The \FUNC{shmem\_malloc\_with\_hints} routine returns a pointer to the allocated space;
    otherwise, it returns a null pointer.
}

	\begin{longtable}{|p{0.45\textwidth}|p{0.5\textwidth}|}
	\hline
	\textbf{Hints} & \textbf{Usage hint}
	\tabularnewline \hline
	\endhead
	%%
	\newline
	\CONST{0} &
	\newline
	Behavior same as \FUNC{shmem\_malloc}
	\tabularnewline \hline
	
		
	\LibConstDecl{SHMEM\_HINT\_ATOMICS\_REMOTE} &
	\newline 
	Memory used for \VAR{atomic} operations
	\tabularnewline \hline
	
	\LibConstDecl{SHMEM\_HINT\_SIGNAL} &
	\newline
	Memory used for \VAR{signal} operations
	\tabularnewline \hline

	\TableCaptionRef{Memory usage hints}
        \label{usagehints}
    \end{longtable}

\apinotes{
		The \openshmem programs should allocate memory with
		\CONST{SHMEM\_HINT\_ATOMICS\_REMOTE}, when the majority of
		operations performed on this memory are atomic operations, and origin
		and target \ac{PE} of the atomic operations do not share a memory domain
		.i.e., symmetric objects on the target \ac{PE} is not accessible using

> 		load/store operations from the origin \ac{PE} or vice versa.

Reviewer 2: Minor: I would consider to move the text under `SHMEM_HINT_ATOMICS_REMOTE` since this is critical information. BTW, probably the same true for remove `SHMEM_HINT_SIGNAL` vs local `SHMEM_HINT_SIGNAL` ? @naveen-rn

Reviewer 3: If I understand the definition of remote vs local wrt this change, I'm not sure whether there is need to differentiate the local side to the signal. All signals from the source PE are passed-by-value to the remote signal address buffer.  A buffer created with SHMEM_HINT_SIGNAL will mostly be updated by some remote PE. 
```bash
void shmem_put_signal(shmem_ctx_t ctx, TYPE *dest, const TYPE *source, size_t nelems,
uint64_t *sig_addr, uint64_t signal, int sig_op, int pe);
```

\apisummary{
    Collective memory allocation routine with support for providing hints.
}

\begin{apidefinition}

\begin{Csynopsis}
void *@\FuncDecl{shmem\_malloc\_with\_hints}@(size_t size, long hints);
\end{Csynopsis}

\begin{apiarguments}
    \apiargument{IN}{size}{The size, in bytes, of a block to be
        allocated from the symmetric heap. This argument is of type \CTYPE{size\_t}}
    \apiargument{IN}{hints}{A bit array of hints provided by the user to the implementation}
\end{apiarguments}

\apidescription{

>     The \FUNC{shmem\_malloc\_with\_hint} like \FUNC{shmem\_malloc} routine returns a pointer to a block of at least

Reviewer 4: ```suggestion
    The \FUNC{shmem\_malloc\_with\_hint} routine, like \FUNC{shmem\_malloc}, returns a pointer to a block of at least
```

    \VAR{size} bytes, which shall be suitably aligned so that it may be
    assigned to a pointer to any type of object.  This space is allocated from
    the symmetric heap (similar to \FUNC{shmem\_malloc}).  When \VAR{size} is zero, 
    the \FUNC{shmem\_malloc\_with\_hint} routine performs no action and returns a null pointer. 
    

>     In addition to the \VAR{size} argument, the \VAR{hint} argument is provided by the user. 

Reviewer 4: ```suggestion
    In addition to the \VAR{size} argument, the \VAR{hints} argument is provided by the user. 
```

>     The \VAR{hint} describes the expected manner in which the \openshmem program may use the allocated memory.

Reviewer 4: ```suggestion
    The \VAR{hints} describes the expected manner in which the \openshmem program may use the allocated memory.
```

>     The valid usage hints are described in Table~\ref{usagehints}. Multiple hints are expressed as \CONST{OR} of \VAR{hints}.

Reviewer 4: ```suggestion
    The valid usage hints are described in Table~\ref{usagehints}. Multiple hints may be requested by combining them with a bitwise \CONST{OR} operation.
```

>     The information provided by the \VAR{hint} is used to optimize for performance by the implementation. 

Reviewer 4: ```suggestion
    The information provided by the \VAR{hints} is used to optimize for performance by the implementation. 
```

    If the implementation cannot optimize, the behavior is same as \FUNC{shmem\_malloc}.
    If more than one hint is provided, the implementation will make the best effort to use one or more hints 
    to optimize performance. 

>             

Reviewer 5: Must all PEs allocate from the same special memory?  It's not clear if asymmetry can exist. Does this impose additional implicit synchronization for each subset configuration of hints if it cannot satisfy the entire hint list?

Also, what happens if you OR `SHMEM_HINT_NONE` with other hint behaviors?

Author: Good question. 

All use cases that I have thought of requires the memory to be symmetric (same kind of memory). 

Regarding extra synchronization, it depends on the implementation. If the implementations maintain asymmetric memory sizes (say each PE starts with different amount of special memory) on the PEs, you might need the extra synchronization for agreement. Otherwise, I do not see a need. In a way, it is similar to current DRAM allocations. Also, for the implementation we explored, we did not need extra synchronization.

I’m reluctant to add such a constraint. Without such constraint, the implementations are free to explore either approach (symmetric and asymmetric). 

Do see value in specifying one way or other?

Reviewer 5: To clarify, what would using `long hint = SHMEM_HINT_NONE | SHMEM_HINT_LOW_LAT_MEM | SHMEM_HINT_HIGH_BW_MEM` do?

Dropping `SHMEM_HINT_NONE`, what if the platform could provide `SHMEM_HINT_LOW_LAT_MEM` or `SHMEM_HINT_HIGH_BW_MEM` but not both simultaneously?  Will we see application code marked up like this because who doesn't want to use low latency and high bandwidth memory for their application? Does the "best effort" default to a platform-specific precedence? The only feedback is that an allocation succeeded or it did not.

Author: > To clarify, what would using long hint = SHMEM_HINT_NONE | SHMEM_HINT_LOW_LAT_MEM | SHMEM_HINT_HIGH_BW_MEM do?

Though this is a legal usage, it does not make sense to use. The implementations are allowed to default to shmem_malloc in this case. 

> Dropping SHMEM_HINT_NONE, what if the platform could provide SHMEM_HINT_LOW_LAT_MEM or SHMEM_HINT_HIGH_BW_MEM but not both simultaneously? Will we see application code marked up like this because who doesn't want to use low latency and high bandwidth memory for their application? Does the "best effort" default to a platform-specific precedence? The only feedback is that an allocation succeeded or it did not.

```
“If more than one hint is provided, the implementation will make the best effort to use one or more hints to optimize performance. 
“
```
My intention with this statement was to provide flexibility for the implementations to optimize as they wish when the user provides multiple hints. Obviously, some combinations of hints might not make sense. In such cases, If the implementations want to give precedence of one hint over others, the proposal allows it. That (assigning priorities to hints) is one way to implement it, but not the only way.

    The \FUNC{shmem\_malloc\_with\_hint} routine is provided  so that multiple \acp{PE} in a program can allocate symmetric,
    remotely accessible memory blocks.  When no action is performed, these

>     routines return without performing a barrier. Otherwise, the routine will call a barrier on exit.

Reviewer 5: > When no action is performed, these routines return without performing a barrier.

What does this mean? That the function returns NULL for all PEs, no memory has been allocated, and no implicit barrier has occurred?  Some applications may use `shmem_malloc` and friends as implicit barriers.  This proposal is for an optimization which seems to be a drop in replacement for vanilla `shmem_malloc`, but dropping the implicit barrier on function exit would change the behavior of the application.

Author: When no allocation is done, “dropping the implicit barrier” is the behavior we have for shmem_malloc in OpenSHMEM 1.4 - please refer page 26 line 40-41. The proposal is aiming to maintain the same behavior for that case.

Reviewer 5: Oops.  In OpenSHMEM 1.4, that was not the case.  This behavior was changed between 1.4 and now in #201.

Author: Ah… Thanks for correcting that.  It has been so long that we debated about this I did not realize it is relatively new. I was looking at my git copy. :) 

Can we consider this issue resolved?

    This ensures that all \acp{PE} participate in the memory allocation, and that the memory on other
    \acp{PE} can be used as soon as the local \ac{PE} returns. The implicit barrier performed by this routine will quiet the
    default context.  It is the user's responsibility to ensure that no communication operations involving the given memory block are pending on
    other contexts prior to calling the \FUNC{shmem\_free} and \FUNC{shmem\_realloc} routines.
    The user is also responsible for calling these routines with identical argument(s) on all

>     \acp{PE}; if differing \VAR{size}, or \VAR{hint} arguments are used, the behavior of the call

Reviewer 4: ```suggestion
    \acp{PE}; if differing \VAR{size}, or \VAR{hints} arguments are used, the behavior of the call
```

    and any subsequent \openshmem calls is undefined.
}

\apireturnvalues{
    The \FUNC{shmem\_malloc\_with\_hint} routine returns a pointer to the allocated space;
    otherwise, it returns a null pointer.
}

\apinotes{ 
}

\apiimpnotes{
}
	\begin{longtable}{|p{0.45\textwidth}|p{0.5\textwidth}|}
	\hline
	\textbf{Hints} & \textbf{Usage hint}
	\tabularnewline \hline
	\endhead
	%%
	\LibConstDecl{SHMEM\_HINT\_NONE} &

> 	Behavior same as shmem\_malloc

Reviewer 4: ```suggestion
	Behavior same as \FUNC{shmem\_malloc}
```

	\tabularnewline \hline
		
	\LibConstDecl{SHMEM\_HINT\_LOW\_LAT\_MEM} &

> 	Allocate memory on low-latency storage	

Reviewer 2: Low-latency relatively to what ? Low latency for local access ? remote access ? both ? only one of those ? Why I would ever what to allocate memory with low-latency and (related to next flag) slow bandwidth ?

Author: > Low-latency relatively to what ? Low latency for local access ? remote access ? both ?

This is symmetric heap’s memory. So, it is for the memory accessed via shmem routines (which I would guess will be dominated by remote access).

Author: > Why I would ever what to allocate memory with low-latency and (related to next flag) slow bandwidth ?

Already discussed here. Please refer to this discussion (https://github.com/openshmem-org/specification/pull/259#discussion_r320468747)

Reviewer 2: The memory is access locally using load and store semantics without shmem. This is how you initialize  the memory with some data in a first place. It is also accessed remotely using load and store semantics without going through SHMEM API.  My point the we cannot claim that something is low latency (the same true for high BW and most of other hints) since it really depends on who is accessing the data and this runtime information. For example, for Nic it can be some local sram region on the die and for the socket it L1/L2/L3. If you want to allocate memory on the nic for low latency access, just call it NIC memory, GPU memory, L1 memory ,etc. The low latency and high bw definitions are misleading.

	\tabularnewline \hline
	
	\LibConstDecl{SHMEM\_HINT\_HIGH\_BW\_MEM} &

> 	Allocate memory on high-bandwidth storage	

Reviewer 6: To capture my preference: "Allocate memory with high bandwidth for local access by the calling PE"

	\tabularnewline \hline

> 	\LibConstDecl{SHMEM\_HINT\_PSYNC} &

Reviewer 2: why we introducing tuning for something that we are deprecating ?

Author: This is always a challenge with tickets (that are developed in parallel) that has impact on each other. 

Already discussed here. Please refer to this discussion (https://github.com/openshmem-org/specification/pull/259#issuecomment-530136683)

	Memory used as \CONST{PSYNC} array

> 	\tabularnewline \hline

Reviewer 2: why we introducing tuning for something that we are deprecating ?

Author: Already discussed here. Please refer to this discussion (https://github.com/openshmem-org/specification/pull/259#issuecomment-530136683)

		
	\LibConstDecl{SHMEM\_HINT\_PWORK} &
	Memory used as \CONST{PWORK} array
	\tabularnewline \hline
	

> 	\LibConstDecl{SHMEM\_HINT\_ATOMICS} &

Reviewer 2: What happens with ATOMIC hint was not passed but atomic operation was used ?

Author: I don't explicitly cover the bad behavior semantics. But, based on the use case experience we have currently, I'm inclined to say that It should not impact correctness. However, the programs might not be able to get the best performance characteristics for that environment.

Do you prefer that we should explicitly say this?

Reviewer 2: My initial assumption is that if this is hint, implementation can completely ignore it. But we still have issue with this semantics. For HCA optimal location of the memory for Atomics will be on the HCA. For core it will be optimal to have in L1/L2 which is suboptimal for ATOMICs coming from the network and vice-versa. Essentially such hint may actually do more  harm to applications. My suggestion is to use explicit names - HCA1, HCA2, GPU, L2, LLC, etc. This will provide very clear description to user what actually happens.

Reviewer 2: Since in this case you optimize for remote PEs only, I think it should be named SHMEM\_HINT\_REMOTE\_ATOMICS to be clear that you don't optimize for local PEs. Without explicit specification local/remote the hint is not very useful.

	Memory used for \VAR{Atomic} operations
	\tabularnewline \hline
	
	\LibConstDecl{SHMEM\_HINT\_SIGNAL} &

> 	Memory used for \VAR{signal} operations

Reviewer 2: What happens when the flag was not used, but signal operation was used ?

Author: Hmm .. Which flag are you referring to, can you please elaborate?

Reviewer 2: Above `SHMEM_HINT_SIGNAL`

	\tabularnewline \hline

	\TableCaptionRef{Memory usage hints}
        \label{usagehints}

>     \end{longtable}

Reviewer 2: Also we should to state that neither alignment requirements or memory properties, such as cache line size are not get impacted by the hint. We also shell state the memory semantics for local assess (load and store) are not impacted by the hint.

Author: AFAIU, for the shmem_malloc routine the implementation is free to allocate the memory, which is either cache aligned or not. One of the constraints is that it should be word-aligned. Similarly, the memory access model (which is yet be defined or clarified here https://github.com/openshmem-org/specification/issues/229) will provide certain access guarantees to the memory allocated by shmem_malloc. In both cases, the proposal intends to follow the semantics of shmem_malloc.

Reviewer 2: The memory gets eventually mapped into the core. The core architecture defined multiple way how it can be mapped. Each on of the mapping has own semantics and constrains. For example semantics between normal cacheable (write back) and non-cacheable (WC) is very different. Since user has direct access through the pointer, code that worked on one machine will break on another with exception or even data corruption.

Reviewer 2: In order to support something like this, you have to remove shmem_ptr and prohibit any direct asses to the memory through load and store semantics. Next you would have to introduce shmem_memcpy function to copy-in-out shmem_malloced region.
